File: radar-auth/src/test/java/org/radarbase/auth/util/TokenTestUtils.java
package org.radarbase.auth.util;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.DecodedJWT;
import com.auth0.jwt.interfaces.RSAKeyProvider;
import org.radarbase.auth.authorization.Permission;

import java.io.IOException;
import java.io.InputStream;
import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.security.cert.Certificate;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.time.Instant;
import java.util.Base64;
import java.util.Date;

/**
 * Sets up a keypair for signing the tokens, initialize all kinds of different tokens for tests.
 */
public final class TokenTestUtils {
    public static final String PUBLIC_KEY = "/oauth/token_key";
    public static final String PUBLIC_KEY_BODY;
    public static final String VALID_RSA_TOKEN;
    public static final String INCORRECT_AUDIENCE_TOKEN;
    public static final String EXPIRED_TOKEN;
    public static final String INCORRECT_ALGORITHM_TOKEN;
    public static final DecodedJWT SCOPE_TOKEN;
    public static final DecodedJWT PROJECT_ADMIN_TOKEN;
    public static final DecodedJWT SUPER_USER_TOKEN;
    public static final DecodedJWT MULTIPLE_ROLES_IN_PROJECT_TOKEN;

    public static final String[] AUTHORITIES = {"ROLE_SYS_ADMIN", "ROLE_USER"};
    public static final String[] ALL_SCOPES = allScopes();
    public static final String[] ROLES = {"PROJECT1:ROLE_PROJECT_ADMIN",
            "PROJECT2:ROLE_PARTICIPANT"};
    public static final String[] SOURCES = {};
    public static final String CLIENT = "unit_test";
    public static final String USER = "admin";
    public static final String ISS = "RADAR";
    public static final String JTI = "some-jwt-id";
    public static final String PUBLIC_KEY_STRING;

    public static final String APPLICATION_JSON = "application/json";
    public static final int WIREMOCK_PORT = 8089;

    private TokenTestUtils() {
        // utility class
    }

    static {
        RSAKeyProvider provider;
        try {
            provider = loadKeys();
        } catch (GeneralSecurityException | IOException e) {
            throw new IllegalStateException("Failed to load keys for test", e);
        }
        RSAPublicKey publicKey = provider.getPublicKeyById("selfsigned");
        PUBLIC_KEY_STRING = new String(Base64.getEncoder().encode(
                provider.getPublicKeyById("selfsigned").getEncoded()));
        Algorithm algorithm = Algorithm.RSA256(publicKey, provider.getPrivateKey());

        PUBLIC_KEY_BODY = "{\n \"keys\" : [ {\n  \"alg\" : \"" + algorithm.getName()
                + "\",\n  \"alg\" : \"RSA\",\n"
                + "  \"value\" : \"-----BEGIN PUBLIC KEY-----\\n" + PUBLIC_KEY_STRING
                + "\\n-----END PUBLIC KEY-----\"\n} ]\n}";

        Instant exp = Instant.now().plusSeconds(30 * 60);
        Instant iat = Instant.now();

        VALID_RSA_TOKEN = initValidToken(algorithm, exp, iat);
        SUPER_USER_TOKEN = JWT.decode(VALID_RSA_TOKEN);
        PROJECT_ADMIN_TOKEN = initProjectAdminToken(algorithm, exp, iat);
        MULTIPLE_ROLES_IN_PROJECT_TOKEN = initMultipleRolesToken(algorithm, exp, iat);
        INCORRECT_AUDIENCE_TOKEN = initIncorrectAudienceToken(algorithm, exp, iat);
        SCOPE_TOKEN = initTokenWithScopes(algorithm, exp, iat);
        INCORRECT_ALGORITHM_TOKEN = initIncorrectAlgorithmToken(exp, iat);

        Instant past = Instant.now().minusSeconds(1);
        Instant iatpast = Instant.now().minusSeconds(30 * 60 + 1);
        EXPIRED_TOKEN = initExpiredToken(algorithm, past, iatpast);
    }

    private static RSAKeyProvider loadKeys() throws GeneralSecurityException, IOException {
        KeyStore ks = KeyStore.getInstance("PKCS12");
        try (InputStream keyStream = Thread.currentThread().getContextClassLoader()
                .getResourceAsStream("keystore.p12")) {
            ks.load(keyStream, "radarbase".toCharArray());
        }
        RSAPrivateKey privateKey = (RSAPrivateKey) ks.getKey("selfsigned",
                "radarbase".toCharArray());
        Certificate cert = ks.getCertificate("selfsigned");
        return new RSAKeyProvider() {
            @Override
            public RSAPublicKey getPublicKeyById(String keyId) {
                return (RSAPublicKey) cert.getPublicKey();
            }

            @Override
            public RSAPrivateKey getPrivateKey() {
                return privateKey;
            }

            @Override
            public String getPrivateKeyId() {
                return null;
            }
        };
    }

    private static String initExpiredToken(Algorithm algorithm, Instant past, Instant iatpast) {
        return JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(Date.from(iatpast))
            .withExpiresAt(Date.from(past))
            .withAudience(CLIENT)
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim("authorities", AUTHORITIES)
            .withArrayClaim("roles", ROLES)
            .withArrayClaim("sources", SOURCES)
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(algorithm);
    }

    private static String initIncorrectAlgorithmToken(Instant exp, Instant iat) {
        Algorithm psk = Algorithm.HMAC256("super-secret-stuff");
        // token signed with a pre-shared key
        return JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(Date.from(iat))
            .withExpiresAt(Date.from(exp))
            .withAudience(CLIENT)
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim("authorities", new String[] {"ROLE_PROJECT_ADMIN"})
            .withArrayClaim("roles", ROLES)
            .withArrayClaim("sources", new String[] {})
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(psk);
    }

    private static String initIncorrectAudienceToken(Algorithm algorithm, Instant exp,
            Instant iat) {
        return JWT.create()
                .withIssuer(ISS)
                .withIssuedAt(Date.from(iat))
                .withExpiresAt(Date.from(exp))
                .withAudience("SOME_AUDIENCE")
                .withSubject(USER)
                .withArrayClaim("scope", ALL_SCOPES)
                .withArrayClaim("authorities", new String[] {"ROLE_PROJECT_ADMIN"})
                .withArrayClaim("roles", ROLES)
                .withArrayClaim("sources", new String[] {})
                .withClaim("client_id", CLIENT)
                .withClaim("user_name", USER)
                .withClaim("jti", JTI)
                .withClaim("grant_type", "password")
                .sign(algorithm);
    }

    private static DecodedJWT initMultipleRolesToken(Algorithm algorithm, Instant exp,
            Instant iat) {
        String multipleRolesInProjectToken = JWT.create()
                .withIssuer(ISS)
                .withIssuedAt(Date.from(iat))
                .withExpiresAt(Date.from(exp))
                .withAudience(CLIENT)
                .withSubject(USER)
                .withArrayClaim("scope", ALL_SCOPES)
                .withArrayClaim("authorities", new String[] {"ROLE_PROJECT_ADMIN"})
                .withArrayClaim("roles", new String[] {"PROJECT2:ROLE_PROJECT_ADMIN",
                        "PROJECT2:ROLE_PARTICIPANT"})
                .withArrayClaim("sources", new String[] {"source-1"})
                .withClaim("client_id", CLIENT)
                .withClaim("user_name", USER)
                .withClaim("jti", JTI)
                .withClaim("grant_type", "password")
                .sign(algorithm);

        return JWT.decode(multipleRolesInProjectToken);
    }

    private static DecodedJWT initProjectAdminToken(Algorithm algorithm, Instant exp, Instant iat) {
        String projectAdminToken = JWT.create()
                .withIssuer(ISS)
                .withIssuedAt(Date.from(iat))
                .withExpiresAt(Date.from(exp))
                .withAudience(CLIENT)
                .withSubject(USER)
                .withArrayClaim("scope", ALL_SCOPES)
                .withArrayClaim("authorities", new String[] {"ROLE_PROJECT_ADMIN",
                        "ROLE_PARTICIPANT"})
                .withArrayClaim("roles", ROLES)
                .withArrayClaim("sources", new String[] {})
                .withClaim("client_id", CLIENT)
                .withClaim("user_name", USER)
                .withClaim("jti", JTI)
                .withClaim("grant_type", "password")
                .sign(algorithm);

        return JWT.decode(projectAdminToken);
    }

    private static String initValidToken(Algorithm algorithm, Instant exp, Instant iat) {
        return JWT.create()
                .withIssuer(ISS)
                .withIssuedAt(Date.from(iat))
                .withExpiresAt(Date.from(exp))
                .withAudience(CLIENT)
                .withSubject(USER)
                .withArrayClaim("scope", ALL_SCOPES)
                .withArrayClaim("authorities", AUTHORITIES)
                .withArrayClaim("roles", ROLES)
                .withArrayClaim("sources", SOURCES)
                .withClaim("client_id", CLIENT)
                .withClaim("user_name", USER)
                .withClaim("jti", JTI)
                .withClaim("grant_type", "password")
                .sign(algorithm);
    }

    private static DecodedJWT initTokenWithScopes(Algorithm algorithm, Instant exp, Instant iat) {
        String token = JWT.create()
                .withIssuer(ISS)
                .withIssuedAt(Date.from(iat))
                .withExpiresAt(Date.from(exp))
                .withAudience(CLIENT)
                .withSubject("i'm a trusted oauth client")
                .withArrayClaim("scope", new String[] {"PROJECT.READ", "SUBJECT.CREATE",
                        "SUBJECT.READ", "MEASUREMENT.CREATE"})
                .withClaim("client_id", "i'm a trusted oauth client")
                .withClaim("jti", JTI)
                .withClaim("grant_type", "client_credentials")
                .sign(algorithm);
        return JWT.decode(token);
    }

    private static String[] allScopes() {
        return Permission.stream()
                .map(Permission::scopeName)
                .toArray(String[]::new);
    }
}

File: radar-auth/src/test/java/org/radarbase/auth/util/TokenTestUtils.kt
package org.radarbase.auth.util

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.interfaces.DecodedJWT
import com.auth0.jwt.interfaces.RSAKeyProvider
import org.radarbase.auth.authorization.Permission
import org.slf4j.LoggerFactory
import java.io.IOException
import java.security.GeneralSecurityException
import java.security.KeyStore
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import java.time.Duration
import java.time.Instant
import java.util.*

/**
 * Sets up a keypair for signing the tokens, initialize all kinds of different tokens for tests.
 */
object TokenTestUtils {
    private val logger = LoggerFactory.getLogger(TokenTestUtils::class.java)

    const val PUBLIC_KEY_PATH = "/oauth/token_key"
    val PUBLIC_KEY_STRING: String
    @JvmField val PUBLIC_KEY_BODY: String
    @JvmField val VALID_RSA_TOKEN: String
    @JvmField val INCORRECT_AUDIENCE_TOKEN: String
    @JvmField val EXPIRED_TOKEN: String
    @JvmField val INCORRECT_ALGORITHM_TOKEN: String
    @JvmField val SCOPE_TOKEN: DecodedJWT
    @JvmField val ORGANIZATION_ADMIN_TOKEN: DecodedJWT
    @JvmField val PROJECT_ADMIN_TOKEN: DecodedJWT
    @JvmField val SUPER_USER_TOKEN: DecodedJWT
    @JvmField val MULTIPLE_ROLES_IN_PROJECT_TOKEN: DecodedJWT
    val AUTHORITIES = arrayOf("ROLE_SYS_ADMIN", "ROLE_USER")
    val ALL_SCOPES = Permission.scopes()
    val ROLES = arrayOf(
        "PROJECT1:ROLE_PROJECT_ADMIN",
        "PROJECT2:ROLE_PARTICIPANT"
    )
    val SOURCES = arrayOf<String>()
    const val CLIENT = "unit_test"
    const val USER = "admin"
    const val ISS = "RADAR"
    const val JTI = "some-jwt-id"
    const val APPLICATION_JSON = "application/json"
    const val WIREMOCK_PORT = 8089

    init {
        val provider: RSAKeyProvider = try {
            loadKeys()
        } catch (e: GeneralSecurityException) {
            throw IllegalStateException("Failed to load keys for test", e)
        } catch (e: IOException) {
            throw IllegalStateException("Failed to load keys for test", e)
        }
        val publicKey = provider.getPublicKeyById("selfsigned")
        PUBLIC_KEY_STRING = String(
            Base64.getEncoder().encode(
                provider.getPublicKeyById("selfsigned").encoded
            )
        )
        val algorithm = Algorithm.RSA256(publicKey, provider.privateKey)
        PUBLIC_KEY_BODY = """
            {
              "keys" : [ {
                "alg" : "${algorithm.name}",
                "kty" : "RSA",
                "value" : "-----BEGIN PUBLIC KEY-----\n$PUBLIC_KEY_STRING\n-----END PUBLIC KEY-----"
              } ]
            }
            """.trimIndent()
        logger.info("Key body {}", PUBLIC_KEY_BODY)
        val exp = Instant.now() + Duration.ofMinutes(30)
        val iat = Instant.now()
        VALID_RSA_TOKEN = initValidToken(algorithm, exp, iat)
        SUPER_USER_TOKEN = JWT.decode(VALID_RSA_TOKEN)
        PROJECT_ADMIN_TOKEN = initProjectAdminToken(algorithm, exp, iat)
        ORGANIZATION_ADMIN_TOKEN = initOrgananizationAdminToken(algorithm, exp, iat)
        MULTIPLE_ROLES_IN_PROJECT_TOKEN = initMultipleRolesToken(algorithm, exp, iat)
        INCORRECT_AUDIENCE_TOKEN = initIncorrectAudienceToken(algorithm, exp, iat)
        SCOPE_TOKEN = initTokenWithScopes(algorithm, exp, iat)
        INCORRECT_ALGORITHM_TOKEN = initIncorrectAlgorithmToken(exp, iat)
        val past = Instant.now().minusSeconds(1)
        val iatpast = Instant.now().minusSeconds((30 * 60 + 1).toLong())
        EXPIRED_TOKEN = initExpiredToken(algorithm, past, iatpast)
    }

    @Throws(GeneralSecurityException::class, IOException::class)
    private fun loadKeys(): RSAKeyProvider {
        val ks = KeyStore.getInstance("PKCS12")
        Thread.currentThread().contextClassLoader
            .getResourceAsStream("keystore.p12")
            .use { keyStream -> ks.load(keyStream, "radarbase".toCharArray()) }
        val privateKey = ks.getKey(
            "selfsigned",
            "radarbase".toCharArray()
        ) as RSAPrivateKey
        val cert = ks.getCertificate("selfsigned")
        return object : RSAKeyProvider {
            override fun getPublicKeyById(keyId: String): RSAPublicKey =
                cert.publicKey as RSAPublicKey

            override fun getPrivateKey(): RSAPrivateKey = privateKey

            override fun getPrivateKeyId(): String = "1"
        }
    }

    private fun initExpiredToken(algorithm: Algorithm, past: Instant, iatpast: Instant): String {
        return JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(iatpast)
            .withExpiresAt(past)
            .withAudience(CLIENT)
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim("authorities", AUTHORITIES)
            .withArrayClaim("roles", ROLES)
            .withArrayClaim("sources", SOURCES)
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(algorithm)
    }

    private fun initIncorrectAlgorithmToken(exp: Instant, iat: Instant): String {
        val psk = Algorithm.HMAC256("super-secret-stuff")
        // token signed with a pre-shared key
        return JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(iat)
            .withExpiresAt(exp)
            .withAudience(CLIENT)
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim("authorities", arrayOf("ROLE_PROJECT_ADMIN"))
            .withArrayClaim("roles", ROLES)
            .withArrayClaim("sources", arrayOf<String>())
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(psk)
    }

    private fun initIncorrectAudienceToken(
        algorithm: Algorithm, exp: Instant,
        iat: Instant
    ): String {
        return JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(iat)
            .withExpiresAt(exp)
            .withAudience("SOME_AUDIENCE")
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim("authorities", arrayOf("ROLE_PROJECT_ADMIN"))
            .withArrayClaim("roles", ROLES)
            .withArrayClaim("sources", arrayOf<String>())
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(algorithm)
    }

    private fun initMultipleRolesToken(
        algorithm: Algorithm, exp: Instant,
        iat: Instant
    ): DecodedJWT {
        val multipleRolesInProjectToken = JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(iat)
            .withExpiresAt(exp)
            .withAudience(CLIENT)
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim("authorities", arrayOf("ROLE_PROJECT_ADMIN"))
            .withArrayClaim(
                "roles", arrayOf(
                    "PROJECT2:ROLE_PROJECT_ADMIN",
                    "PROJECT2:ROLE_PARTICIPANT"
                )
            )
            .withArrayClaim("sources", arrayOf("source-1"))
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(algorithm)
        return JWT.decode(multipleRolesInProjectToken)
    }

    private fun initOrgananizationAdminToken(
        algorithm: Algorithm, exp: Instant,
        iat: Instant
    ): DecodedJWT {
        val projectAdminToken = JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(iat)
            .withExpiresAt(exp)
            .withAudience(CLIENT)
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim("authorities", arrayOf("ROLE_ORGANIZATION_ADMIN"))
            .withArrayClaim("roles", arrayOf("main:ROLE_ORGANIZATION_ADMIN"))
            .withArrayClaim("sources", arrayOf<String>())
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(algorithm)
        return JWT.decode(projectAdminToken)
    }

    private fun initProjectAdminToken(
        algorithm: Algorithm,
        exp: Instant,
        iat: Instant
    ): DecodedJWT {
        val projectAdminToken = JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(iat)
            .withExpiresAt(exp)
            .withAudience(CLIENT)
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim(
                "authorities", arrayOf(
                    "ROLE_PROJECT_ADMIN",
                    "ROLE_PARTICIPANT"
                )
            )
            .withArrayClaim("roles", ROLES)
            .withArrayClaim("sources", arrayOf<String>())
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(algorithm)
        return JWT.decode(projectAdminToken)
    }

    private fun initValidToken(algorithm: Algorithm, exp: Instant, iat: Instant): String {
        return JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(iat)
            .withExpiresAt(exp)
            .withAudience(CLIENT)
            .withSubject(USER)
            .withArrayClaim("scope", ALL_SCOPES)
            .withArrayClaim("authorities", AUTHORITIES)
            .withArrayClaim("roles", ROLES)
            .withArrayClaim("sources", SOURCES)
            .withClaim("client_id", CLIENT)
            .withClaim("user_name", USER)
            .withClaim("jti", JTI)
            .withClaim("grant_type", "password")
            .sign(algorithm)
    }

    private fun initTokenWithScopes(algorithm: Algorithm, exp: Instant, iat: Instant): DecodedJWT {
        val token = JWT.create()
            .withIssuer(ISS)
            .withIssuedAt(iat)
            .withExpiresAt(exp)
            .withAudience(CLIENT)
            .withSubject("i'm a trusted oauth client")
            .withArrayClaim(
                "scope", arrayOf(
                    "PROJECT.READ", "SUBJECT.CREATE",
                    "SUBJECT.READ", "MEASUREMENT.CREATE"
                )
            )
            .withClaim("client_id", "i'm a trusted oauth client")
            .withClaim("jti", JTI)
            .withClaim("grant_type", "client_credentials")
            .sign(algorithm)
        return JWT.decode(token)
    }
}
